File: sdk/v1.0.15/python/mediaviz_sdk/keywords.py
from __future__ import annotations
from typing import Any
from urllib.parse import quote, urlencode
import httpx

from .errors import handle_response


class Keywords:
    def __init__(self, ctx) -> None:
        self._ctx = ctx

    def create_keyword_filtering_list(
        self,
        name: str,
        projectList: list | None = None,
    ) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/keyword/'
        body = {k: v for k, v in {
            'name': name,
            'project_list': projectList,
        }.items() if v is not None}
        return self._ctx.client.request(path, 'POST', self._ctx.access_token, self._ctx.refresh_token, body).data

    def get_user_keyword_filtering_lists(self) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/keyword/user'
        return self._ctx.client.request(path, 'GET', self._ctx.access_token, self._ctx.refresh_token).data

    def get_keyword_filtering_list_and_projects_by_id(self, keyword_list_id: int) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/keyword/' + quote(str(keyword_list_id), safe='')
        return self._ctx.client.request(path, 'GET', self._ctx.access_token, self._ctx.refresh_token).data

    def get_keyword_filtering_list_by_id(self, keyword_list_id: int) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/keyword/list/' + quote(str(keyword_list_id), safe='')
        return self._ctx.client.request(path, 'GET', self._ctx.access_token, self._ctx.refresh_token).data

    def get_existing_keyword_filtering_list_by_project(
        self,
        project_table_name: str,
    ) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/keyword/project/' + quote(str(project_table_name), safe='')
        return self._ctx.client.request(path, 'GET', self._ctx.access_token, self._ctx.refresh_token).data

    def get_default_keyword_filtering_list_by_project(
        self,
        project_table_name: str,
    ) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/keyword/project/' + quote(str(project_table_name), safe='') + '/default'
        return self._ctx.client.request(path, 'GET', self._ctx.access_token, self._ctx.refresh_token).data

    def update_keyword_filtering_list_labels(
        self,
        keyword_list_id: int,
        listKeywordsToInclude: list,
        listKeywordsToExclude: list,
    ) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/keyword/' + quote(str(keyword_list_id), safe='')
        body = {k: v for k, v in {
            'list_keywords_to_include': listKeywordsToInclude,
            'list_keywords_to_exclude': listKeywordsToExclude,
        }.items() if v is not None}
        return self._ctx.client.request(path, 'PUT', self._ctx.access_token, self._ctx.refresh_token, body).data

    def update_keyword_filtering_list_details(
        self,
        keyword_list_id: int,
        name: str | None = None,
        projectList: list | None = None,
    ) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/keyword/details/' + quote(str(keyword_list_id), safe='')
        body = {k: v for k, v in {
            'name': name,
            'project_list': projectList,
        }.items() if v is not None}
        return self._ctx.client.request(path, 'PUT', self._ctx.access_token, self._ctx.refresh_token, body).data

    def add_project_to_keyword_filtering_list(
        self,
        keyword_list_id: int,
        project_ids: Any | None = None,
    ) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/keyword/' + quote(str(keyword_list_id), safe='') + '/projects'
        _q: dict[str, str] = {}
        if project_ids is not None:
            _q['project_ids'] = str(project_ids)
        if _q:
            path += '?' + urlencode(_q)
        return self._ctx.client.request(path, 'POST', self._ctx.access_token, self._ctx.refresh_token).data

    def request_keyword_list_export(self, keyword_list_id: int) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/keyword/export/' + quote(str(keyword_list_id), safe='')
        return self._ctx.client.request(path, 'GET', self._ctx.access_token, self._ctx.refresh_token).data

    def request_keyword_list_export_status(self, keyword_list_id: int) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/keyword/export_status/' + quote(str(keyword_list_id), safe='')
        return self._ctx.client.request(path, 'GET', self._ctx.access_token, self._ctx.refresh_token).data

    def get_keywords_and_ids(self) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/keyword/all_keywords/id/label'
        return self._ctx.client.request(path, 'GET', self._ctx.access_token, self._ctx.refresh_token).data

    def remove_projects_from_keyword_filtering_list(
        self,
        keyword_list_id: int,
        project_ids: Any | None = None,
    ) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/keyword/' + quote(str(keyword_list_id), safe='') + '/projects'
        _q: dict[str, str] = {}
        if project_ids is not None:
            _q['project_ids'] = str(project_ids)
        if _q:
            path += '?' + urlencode(_q)
        return self._ctx.client.request(path, 'DELETE', self._ctx.access_token, self._ctx.refresh_token).data

    def delete_keyword_filtering_list_by_id(self, keyword_list_id: int) -> dict[str, Any]:
        self._ctx.require_tokens()
        path = '/api/v1/keyword/' + quote(str(keyword_list_id), safe='')
        return self._ctx.client.request(path, 'DELETE', self._ctx.access_token, self._ctx.refresh_token).data

File: sdk/v1.0.19/javascript/keywords.js
function stripUndef(o) { const r = {}; for (const k in o) if (o[k] !== undefined) r[k] = o[k]; return r; }

export class Keywords {
  constructor(ctx) { this._ctx = ctx; }

  async createKeywordFilteringList(name, projectList = undefined) {
    this._ctx.requireTokens();
    const path = `/api/v1/keyword/`;
    const body = stripUndef({
      name: name,
      project_list: projectList,
    });
    const { data } = await this._ctx.client.request(path, 'POST', this._ctx.accessToken, this._ctx.refreshToken, body);
    return data;
  }

  async getUserKeywordFilteringLists() {
    this._ctx.requireTokens();
    const path = `/api/v1/keyword/user`;
    const { data } = await this._ctx.client.request(path, 'GET', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async getKeywordFilteringListAndProjectsById(keywordListId) {
    this._ctx.requireTokens();
    const path = `/api/v1/keyword/${encodeURIComponent(keywordListId)}`;
    const { data } = await this._ctx.client.request(path, 'GET', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async getKeywordFilteringListById(keywordListId) {
    this._ctx.requireTokens();
    const path = `/api/v1/keyword/list/${encodeURIComponent(keywordListId)}`;
    const { data } = await this._ctx.client.request(path, 'GET', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async getExistingKeywordFilteringListByProject(projectTableName) {
    this._ctx.requireTokens();
    const path = `/api/v1/keyword/project/${encodeURIComponent(projectTableName)}`;
    const { data } = await this._ctx.client.request(path, 'GET', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async getDefaultKeywordFilteringListByProject(projectTableName) {
    this._ctx.requireTokens();
    const path = `/api/v1/keyword/project/${encodeURIComponent(projectTableName)}/default`;
    const { data } = await this._ctx.client.request(path, 'GET', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async updateKeywordFilteringListLabels(keywordListId, listKeywordsToInclude, listKeywordsToExclude) {
    this._ctx.requireTokens();
    const path = `/api/v1/keyword/${encodeURIComponent(keywordListId)}`;
    const body = stripUndef({
      list_keywords_to_include: listKeywordsToInclude,
      list_keywords_to_exclude: listKeywordsToExclude,
    });
    const { data } = await this._ctx.client.request(path, 'PUT', this._ctx.accessToken, this._ctx.refreshToken, body);
    return data;
  }

  async updateKeywordFilteringListDetails(keywordListId, name = undefined, projectList = undefined) {
    this._ctx.requireTokens();
    const path = `/api/v1/keyword/details/${encodeURIComponent(keywordListId)}`;
    const body = stripUndef({
      name: name,
      project_list: projectList,
    });
    const { data } = await this._ctx.client.request(path, 'PUT', this._ctx.accessToken, this._ctx.refreshToken, body);
    return data;
  }

  async addProjectToKeywordFilteringList(keywordListId, { projectIds } = {}) {
    this._ctx.requireTokens();
    let path = `/api/v1/keyword/${encodeURIComponent(keywordListId)}/projects`;
    const query = new URLSearchParams();
    if (projectIds !== undefined) (Array.isArray(projectIds) ? projectIds : [projectIds]).forEach(v => query.append('project_ids', v));
    const qs = query.toString();
    if (qs) path += '?' + qs;
    const { data } = await this._ctx.client.request(path, 'POST', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async requestKeywordListExport(keywordListId) {
    this._ctx.requireTokens();
    const path = `/api/v1/keyword/export/${encodeURIComponent(keywordListId)}`;
    const { data } = await this._ctx.client.request(path, 'GET', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async requestKeywordListExportStatus(keywordListId) {
    this._ctx.requireTokens();
    const path = `/api/v1/keyword/export_status/${encodeURIComponent(keywordListId)}`;
    const { data } = await this._ctx.client.request(path, 'GET', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async getKeywordsAndIds() {
    this._ctx.requireTokens();
    const path = `/api/v1/keyword/all_keywords/id/label`;
    const { data } = await this._ctx.client.request(path, 'GET', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async removeProjectsFromKeywordFilteringList(keywordListId, { projectIds } = {}) {
    this._ctx.requireTokens();
    let path = `/api/v1/keyword/${encodeURIComponent(keywordListId)}/projects`;
    const query = new URLSearchParams();
    if (projectIds !== undefined) (Array.isArray(projectIds) ? projectIds : [projectIds]).forEach(v => query.append('project_ids', v));
    const qs = query.toString();
    if (qs) path += '?' + qs;
    const { data } = await this._ctx.client.request(path, 'DELETE', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }

  async deleteKeywordFilteringListById(keywordListId) {
    this._ctx.requireTokens();
    const path = `/api/v1/keyword/${encodeURIComponent(keywordListId)}`;
    const { data } = await this._ctx.client.request(path, 'DELETE', this._ctx.accessToken, this._ctx.refreshToken);
    return data;
  }
}
